azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -223,73 +223,141 @@ def _handle_pipeline_event(self, event):
 
 
 class EnsureConnectionStage(PipelineStage):
-    # TODO: additional documentation and tests for this class are not being implemented because a significant rewriting to support more scenarios is pending
     """
     This stage is responsible for ensuring that the protocol is connected when
-    it needs to be connected, and it's responsible for queueing operations
-    while we're waiting for the connect operation to complete.
-
-    Note: this stage will likely be replaced by a more full-featured stage to handle
-    other "block while we're setting something up" operations, such as subscribing to
-    twin responses.  That is another example where we want to ensure some state and block
-    requests until that state is achieved.
+    it needs to be connected.
     """
 
     def __init__(self):
         super(EnsureConnectionStage, self).__init__()
         self.connected = False
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _execute_op(self, op):
+        # Any operation that requires a connection can trigger a connection if
+        # we're not connected.
+        if op.needs_connection and not self.connected:
+            self._do_connect(op)
+
+        # Finally, if this stage doesn't need to do anything else with this operation,
+        # it just passes it down.
+        else:
+            operation_flow.pass_op_to_next_stage(self, op)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _do_connect(self, op):
+        """
+        Start connecting the transport in response to some operation
+        """
+        # function that gets called after we're connected.
+        @pipeline_thread.runs_on_pipeline_thread
+        def on_connected(op_connect):
+            logger.info(
+                "{}({}): connection is complete: {}".format(self.name, op.name, op_connect.error)
```

Comment:
>op_connect.error [](start = 80, length = 16)

Why would there be an error on connection complete?

Reply:
If the connection failed.  The operation is done, but it failed.  I think i can re-phrase this message better.

---
In reply to: [311697780](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r311697780) [](ancestors = 311697780)